datafusion/optimizer/src/analyzer/type_coercion.rs

```diff
@@ -1428,33 +1404,47 @@ mod test {
     }
 
     #[test]
-    fn in_subquery() -> Result<()> {
-        let empty_inside = Arc::new(LogicalPlan::EmptyRelation(EmptyRelation {
-            produce_one_row: false,
-            schema: Arc::new(DFSchema::new_with_metadata(
-                vec![DFField::new_unqualified("a_int32", DataType::Int32, true)],
-                std::collections::HashMap::new(),
-            )?),
-        }));
-        let empty_outside = Arc::new(LogicalPlan::EmptyRelation(EmptyRelation {
-            produce_one_row: false,
-            schema: Arc::new(DFSchema::new_with_metadata(
-                vec![DFField::new_unqualified("a_int64", DataType::Int64, true)],
-                std::collections::HashMap::new(),
-            )?),
-        }));
+    fn in_subquery_cast_subquery() -> Result<()> {
+        let empty_int32 = empty_with_type(DataType::Int32);
+        let empty_int64 = empty_with_type(DataType::Int64);
+
+        let in_subquery_expr = Expr::InSubquery {
+            expr: Box::new(col("a")),
+            subquery: Subquery {
+                subquery: empty_int32,
+                outer_ref_columns: vec![],
+            },
+            negated: false,
+        };
+        let plan = LogicalPlan::Filter(Filter::try_new(in_subquery_expr, empty_int64)?);
+        // add cast for subquery
+        let expected = "\
+        Filter: a IN (<subquery>)\
+        \n  Subquery:\
+        \n    Projection: CAST(a AS Int64)\
+        \n      EmptyRelation\
+        \n  EmptyRelation";
+        assert_analyzed_plan_eq(Arc::new(TypeCoercion::new()), &plan, expected)?;
+        Ok(())
+    }
+
+    #[test]
+    fn in_subquery_cast_expr() -> Result<()> {
```

Comment:
could you please add or replace the cast case for decimal data type?
the coercion type or the common type is not same with `expr` and `subquery`, this case will both cast `expr` and `subquer` together.